### PR TITLE
Create C# How to Section index

### PR DESCRIPTION
**Pull request by BillWagner**
Over time, the existing How To articles will go in this section as they are updated for more common current scenarios. 

In the meantime, the index provides a list of the existing How tos.

I have a few questions for @mairaw and @rpetrusha as I keep working on this:
- Is there way too much LINQ content? There is a lot of LINQ to XML examples. Are those all still relevant?
- I already removed the How to articles relating to reflection and CodeDom. Do you agree?

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/how-to/index?branch=pr-en-us-4031)

### CODE REVIEW

**Comment by BillWagner** [comment]
This is ready for a review @rpetrusha @mairaw

File: docs/csharp/how-to/index.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+ms.author: wiwagn
+ms.date: 12/20/2017
+ms.topic: article
+ms.prod: .net
+ms.devlang: devlang-csharp
+---
+
+# How to (C#)
+
+The How to section of the C# Guide is where you can find quick answers
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: Better: In the How to section of the C# Guide, you can find...

File: docs/csharp/how-to/index.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+to common questions. There are cases on this page were one article may
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: In some cases, articles are listed in multiple sections on this page. We want...

File: docs/csharp/how-to/index.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+- [How to use lambda expressions](../programming-guide/statements-expressions-operators/how-to-use-lambda-expressions-outside-linq.md).
+- [Resolve type name conflicts by using the global namespace alias](../programming-guide/namespaces/how-to-use-the-global-namespace-alias.md).
+- [Use operator overloading](../programming-guide/statements-expressions-operators/how-to-use-operator-overloading-to-create-a-complex-number-class.md).
+- [Implement and call a custom extension method](../programming-guide/classes-and-structs/how-to-implement-and-call-a-custom-extension-method.md).
+- Even C# programmers may want to [use the `My` namespace from VB](../programming-guide/namespaces/how-to-use-the-my-namespace.md).
+- [Create a new method for an `enum` type using extension methods](../programming-guide/classes-and-structs/how-to-create-a-new-method-for-an-enumeration.md).
+
+### Class and struct members
+
+You create classes and structs to implement your program. These techniques are common tasks when writing classes or structs.
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: techniques and tasks are different things. Maybe "These techniques are commonly used when writing..."

File: docs/csharp/how-to/index.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+- [Declare auto implemented properties](../programming-guide/classes-and-structs/how-to-implement-a-lightweight-class-with-auto-implemented-properties.md).
+- [Declare and use read/write properties](../programming-guide/classes-and-structs/how-to-declare-and-use-read-write-properties.md).
+- [Define constants](../programming-guide/classes-and-structs/how-to-define-constants.md).
+- [Override the `ToString` method to provide string output](../programming-guide/classes-and-structs/how-to-override-the-tostring-method.md).
+- [Define abstract properties](../programming-guide/classes-and-structs/how-to-define-abstract-properties.md).
+- [Use the xml documentation features to document your code](../programming-guide/xmldoc/how-to-use-the-xml-documentation-features.md).
+- [Explicitly implement interface members](../programming-guide/interfaces/how-to-explicitly-implement-interface-members.md) to keep your public interface more clear.
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: "more clear" seems like the wrong phrase. Perhaps just small? Or lean?

File: docs/csharp/how-to/index.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+You may need to convert an object to a different type.
+
+- [Determine if a string represents a number](../programming-guide/strings/how-to-determine-whether-a-string-represents-a-numeric-value.md).
+- [Convert between strings that represent hexadecimal numbers and the number](../programming-guide/types/how-to-convert-between-hexadecimal-strings-and-numeric-types.md).
+- [Convert a string to a <xref:System.DateTime>](../programming-guide/strings/how-to-convert-a-string-to-a-datetime.md).
+- [Convert a byte array to an int](../programming-guide/types/how-to-convert-a-byte-array-to-an-int.md).
+- [Convert a string to a number](../programming-guide/types/how-to-convert-a-string-to-a-number.md).
+- [Use `as` and `is` to safely cast to a different type](../programming-guide/types/how-to-safely-cast-by-using-as-and-is-operators.md).
+- [Define conversion operators for `struct` types](../programming-guide/statements-expressions-operators/how-to-implement-user-defined-conversions-between-structs.md).
+- [Determine if a type is a nullable type](../programming-guide/nullable-types/how-to-identify-a-nullable-type.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: or just "Determine if a type is nullable"?

File: docs/csharp/how-to/index.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+- [Convert between nullable and non-nullable value types](../programming-guide/nullable-types/how-to-safely-cast-from-bool-to-bool.md).
+
+## Equality and ordering comparisons
+
+You may create types that define their own rules for equality or define a
+natural ordering among objects of that type.
+
+- [Test for reference based equality](../programming-guide/statements-expressions-operators/how-to-test-for-reference-equality-identity.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: reference-based

File: docs/csharp/how-to/index.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+- [Define value based equality for a type](../programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: value-based

File: docs/csharp/how-to/index.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+## Exception handling
+
+Throwing exceptions is how .NET programs report that methods did not
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: The "is how" is awkward. Perhaps ".NET programs report that methods did not successfully complete their work by throwing exceptions."

File: docs/csharp/how-to/index.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+successfully complete their work. In these articles you'll learn to work
+with exceptions.
+
+- [Handle exceptions using `try` and `catch`](../programming-guide/exceptions/how-to-handle-an-exception-using-try-catch.md).
+- [Cleanup resources using `finally` clauses](../programming-guide/exceptions/how-to-execute-cleanup-code-using-finally.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: Cleanup --> Clean up (2 words as a verb, 1 as a noun)

File: docs/csharp/how-to/index.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+- [Recover from non-CLS (Common Language Specification) exceptions](../programming-guide/exceptions/how-to-catch-a-non-cls-exception.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: This is a super-nit, but I don't think many people know what a non-CLS exception is, so I wonder if it's worthwhile to change the title. But on the other hand, if you don't know what it is, you probably don't need to handle it.

File: docs/csharp/how-to/index.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+## Delegates and events
+
+Delegates and events provide a capability for strategies that involve
+loosely coupled blocks of code.
+
+- [Declare, instantiate, and use delegates](../programming-guide/delegates/how-to-declare-instantiate-and-use-a-delegate.md).
+- [Combine multicast delegates](../programming-guide/delegates/how-to-combine-delegates-multicast-delegates.md).
+
+Events provide a mechanism to publish notifications or subscribe to notifications.
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: publish notifications or subscribe to notifications --> publish or subscribe to notifications.

File: docs/csharp/how-to/index.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+- [Subscribe and unsubscrbe from events](../programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: unsubscrbe --> unsubscribe

File: docs/csharp/how-to/index.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+- [Work with dictionaries using LINQ to XML](../programming-guide/concepts/linq/how-to-work-with-dictionaries-using-linq-to-xml.md).
+- [Retrieve paragraphs from an office open XML document](../programming-guide/concepts/linq/how-to-retrieve-paragraphs-from-an-office-open-xml-document.md).
+- [Modify an office Open XML document](../programming-guide/concepts/linq/how-to-modify-an-office-open-xml-document.md).
+
+## Multiple threads and async processing
+
+Modern programs often use asynchronous operations. These articles will help you learn
+to use these techniques.
+
+- [Improve async performance using <xref:System.Threading.Tasks.Task.WhenAll%2A>](../programming-guide/concepts/async/how-to-extend-the-async-walkthrough-by-using-task-whenall.md).
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: Links within a link don't resolve, so replace the xref with backticks.

File: docs/csharp/how-to/index.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+- [Make multiple web requests in parallel using `async` and `await`](../programming-guide/concepts/async/how-to-make-multiple-web-requests-in-parallel-by-using-async-and-await.md).
+- [Use a thread pool](../programming-guide/concepts/threading/how-to-use-a-thread-pool.md).
+
+## Command line args to your program
+
+Typically, C# programs have command line arguments. These articles teach you to access and process
+those command line arguments.
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: Although both of the topics are about displaying command line arguments, I'd change the description here to retrieving them, which strikes me as a much more common operation than displaying them.

File: docs/csharp/how-to/index.md [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+Events provide a mechanism to publish notifications or subscribe to notifications.
+
+- [Subscribe and unsubscrbe from events](../programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events.md).
+- [Implement events declared in interfaces](../programming-guide/events/how-to-implement-interface-events.md).
+- [Conform to .NET Framework guidelines when your code publishes events](../programming-guide/events/how-to-publish-events-that-conform-to-net-framework-guidelines.md).
+- [Raise events defined in base classes from derived classes](../programming-guide/events/how-to-raise-base-class-events-in-derived-classes.md).
+- [Store event instances in a dictionary](../programming-guide/events/how-to-use-a-dictionary-to-store-event-instances.md).
+- [Implement custom event accessors](../programming-guide/events/how-to-implement-custom-event-accessors.md).
+
+## LINQ practices
```
Comment (rpetrusha) [inline docs/csharp/how-to/index.md]: Yes, there really is too much LINQ content; the LINQ section really overshadows everything else. Most are also not so much language topics as LINQ topics.   I think that I would pare this down to a few of the more basic topics in each category, and possibly plan to create a more thorough list of LINQ how to topics in the LINQ docs later.

**Review by rpetrusha — CHANGES_REQUESTED** [review]
I left a number of comments and suggestions, @BillWagner. I agree that the CodeDom articles can go. And I think that there is way too much LINQ content -- I suggested shortening the list to a few more simple topics per category, and eventually adding a LINQ how-to topics list in the LINQ docs. In terms of reflection, I'm somewhat mixed. But given that it's a purely framework and not a language topic, that seems to me to argue for its exclusion.

File: docs/csharp/how-to/index.md (hunk 4 above)
Comment (BillWagner) [reply docs/csharp/how-to/index.md]: How about "concise"? I'd like to capture both size and clarity.

File: docs/csharp/how-to/index.md (hunk 5 above)
Comment (BillWagner) [reply docs/csharp/how-to/index.md]: I changed to nullable value type.  I want to future-proof this against the upcoming non-nullable reference types. (The referenced article doesn't help determine nullable vs. non-nullable reference types, only value types.)

File: docs/csharp/how-to/index.md (hunk 10 above)
Comment (BillWagner) [reply docs/csharp/how-to/index.md]: Overall on this article, I took the 2nd approach: The How-To articles are meant to solve a specific problem, and the reader should be familiar with enough of the terms for their particular problem.

File: docs/csharp/how-to/index.md (hunk 15 above)
Comment (BillWagner) [reply docs/csharp/how-to/index.md]: I pared it down to the first general How-tos. Those focus more on the query syntax, which is closer to the language.

**Comment by BillWagner** [comment]
@rpetrusha Thanks for the review.  I updated this based on your feedback. In a couple spots I chose different wording than your proposed changes, and commented on those changes.

Let me know what you think on this round.

**Review by rpetrusha — APPROVED** [review]

**Comment by rpetrusha** [comment]
I think that this looks really good, @BillWagner, especially the pared-down LINQ section. It's ready to merge when you're ready.

**Pull request merged**